---
title: "Covariate Analysis"
author: "Thanneer Perumal"
date: "02/20/2015"
output: html_document
---

```{r libs, echo=FALSE, warning=FALSE, message=FALSE}
## It is assumed your working directory is where this file is

# Load required libraries
library('synapseClient')
library('limma')
library('edgeR')
library('RColorBrewer')
library('ctv')
library('ggplot2')
library('psych')
library('reshape2')
library('gplots')
library('vcd')
library('psych')

## Requires ggdendro
devtools::source_url("https://raw.githubusercontent.com/chr1swallace/random-functions/master/R/ggplot-heatmap.R")

synapseLogin()

# source utility files from ../R/lib folder
file.sources = list.files('../R/lib',pattern="*.R")
file.sources = sapply(file.sources,function(x){return(paste('../R/lib',x,sep='/'))})
tmp = sapply(file.sources,source,.GlobalEnv)

options(xtable.type="html")

knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE)
```

```{r setup}
# Input Parameters
COUNT_ID = 'syn3164570';
METADATA_ID = 'syn3156503';

# Specify factor and continuous covarites pool (adjusted covariates will only be the subset of these covariates)
FactorCovariates = c('Diffname_short','run','lane','index','Cell_Type_of_Origin',
                     'Tissue_of_Origin', 'Reprogramming_Vector_Type','Reprogramming_Gene_Combination',
                     'Donor_Life_Stage','Originating_Lab','Gender','High_Confidence_Donor_ID')
ContCovariates = c('numberReads','mappedReads')

print(paste('Factor covariates considered for analysis are',paste(FactorCovariates,collapse=','),', and continuous covariates considered for the analysis are',paste(ContCovariates,collapse=','),sep=' '))
```

```{r getdata, cache=TRUE}
# Get mRNA counts and metadata table

# Get count matrix
COUNT_OBJ = synGet(COUNT_ID);
ALL_USED_OBJ = list(COUNT_OBJ);
COUNT = read.table(getFileLocation(COUNT_OBJ),header=T,row.names=1, check.names=FALSE)

# Get metadata
METADATA_OBJ = synTableQuery(paste('SELECT * FROM',METADATA_ID,sep=' '))
ALL_USED_OBJ[length(ALL_USED_OBJ)+1] = METADATA_OBJ;
METADATA = METADATA_OBJ@values

# Preprocess metadata
METADATA[METADATA == 'N/A'] = NA
METADATA$Diffname_short[grep('MESO',METADATA$Diffname_short)] = 'MESO'

# Arrange count and metadata
RowsPresent = match(colnames(COUNT), METADATA$UID);
METADATA = METADATA[RowsPresent,]
rownames(METADATA) = METADATA$UID

#### Pre processing mRNA expression counts and metadata ####
# Remove samples that failed QC
COUNT = COUNT[,METADATA$pass_qc]
METADATA = METADATA[METADATA$pass_qc,]

# Remove samples from bad lines
COUNT = COUNT[,!METADATA$exclude]
METADATA = METADATA[!METADATA$exclude,]

# Remove somatic samples and samples with type N/A
ind = METADATA$Cell_Type == "PSC" & !is.na(METADATA$Cell_Type)
COUNT = COUNT[,ind]
METADATA = METADATA[ind,]

# Remove samples with abnormal karyotypes
ind = which(METADATA$C4_Karyotype_Result != "abnormal" | is.na(METADATA$C4_Karyotype_Result))
COUNT = COUNT[,ind]
METADATA = METADATA[ind,]

```

### Normalisation

Determine design matrix for normalisation and differential expression analysis. Remove genes that have less than 1 cpm counts in at least 50% of samples.

```{r}
PROCESSED_COUNTS = getGeneFilteredGeneExprMatrix(COUNT)
```

```{r CPM_HISTOGRAM}
print(PROCESSED_COUNTS$plotHist)
```

### Covariate clustering

```{r COVARIATES_CLUSTERING, fig.width=12, fig.height=10}

# Find inter relation between factor covariates
COVARIATES = METADATA[,c(FactorCovariates,ContCovariates)]

# Convert factor to numeric matrix 
COVARIATES[,FactorCovariates] = apply(COVARIATES[,FactorCovariates],2,function(cols){cols=as.numeric(unclass(factor(cols)))})

heatmap.2(as.matrix(COVARIATES), col = brewer.pal(11,'Spectral'), scale='column',
          Rowv = F, cex.lab = 1.5, cex.axis = 1.5, dendrogram = 'column',
          labRow=rep(as.character(' '),dim(COVARIATES)[1]),
          cex.main = 1.5, mar=c(18,10), xlab = "Covariates", ylab = "Samples", 
          main = "Clustering of covariates", keysize=1)

```

### Covariate correlation

```{r COVARIATES.CORRELATION, fig.width=12, fig.height=10}
COVARIATES.CORRELATION = getAssociationStatistics(COVARIATES,FactorCovariates,ContCovariates,PVAL=1)
ggheatmap.show(COVARIATES.CORRELATION$plot, col.width=0.3, row.width=0.15)
```

Initial normalisation usign voom (with NULL design)

```{r}
VOOM.GENE_EXPRESSION = voom(PROCESSED_COUNTS$filteredExprMatrix, design=NULL, plot=T)

# Find PC of gene expression and significant covariates that are highly correlated with PCs
DM = getDesignMatrix(METADATA[,c(FactorCovariates,ContCovariates)],FactorCovariates)
```

```{r}
# PC_res=list()
preAdjustedSigCovars = runPCAandPlotCorrelations(VOOM.GENE_EXPRESSION$E, DM$design,
                                                 'NULL design(voom-normalized)', isKeyPlot=TRUE)
```

```{r preAdjustedSigCovars.NULL, fig.width=20, fig.height=10}
preAdjustedSigCovars[["PC_res"]][[1]]$plotData
```

```{r preAdjustedSigCovars.NULL.ALL, fig.width=20, fig.height=10}
preAdjustedSigCovars[["PC_res"]][[2]]$plotData
```

Re-normalise with COVARIATES design matrix.

```{r ADJUSTED_PCA}
adjustCovars = designMatVarsToCovars(getCovariatesMap(DM),preAdjustedSigCovars$significantCovars)
print(paste('Significant covariates to adjust at FDR 0.1 are',paste(adjustCovars,collapse=','),sep=' '))

DM1 = getDesignMatrix(METADATA[,adjustCovars],FactorCovariates)
VOOM.ADJUSTED.GENE_EXPRESSION = voom(PROCESSED_COUNTS$filteredExprMatrix, design=DM1$design, plot=T)

# Find PC of re-adjusted gene expression and significant covariates that are highly correlated with PCs
postAdjustedSigCovars = runPCAandPlotCorrelations(VOOM.ADJUSTED.GENE_EXPRESSION$E, DM$design,
                                                  'all adjusted design(voom-normalized)',isKeyPlot=TRUE)
```

```{r postAdjustedSigCovars, fig.width=20, fig.height=10}
postAdjustedSigCovars[["PC_res"]][[1]]$plotData
```

```{r postAdjustedSigCovars.ALL, fig.width=20, fig.height=10}
postAdjustedSigCovars[["PC_res"]][[2]]$plotData
```

```{r CALCULATE.RESIDUALS}
RESIDUAL.GENE_EXPRESSION = calcResiduals(VOOM.ADJUSTED.GENE_EXPRESSION$E, DM1$design)

# Find PC of re-adjusted gene expression and significant covariates that are highly correlated with PCs
residualSigCovars = runPCAandPlotCorrelations(RESIDUAL.GENE_EXPRESSION, DM$design,
                                                  'residual (voom-normalized)',isKeyPlot=TRUE)
```

```{r residualSigCovars, fig.width=20, fig.height=10}
residualSigCovars[["PC_res"]][[1]]$plotData
```

```{r residualSigCovars.ALL, fig.width=20, fig.height=10}
residualSigCovars[["PC_res"]][[2]]$plotData
```